cmd/gateway-azure.go

```diff
@@ -457,70 +497,58 @@ func (a *azureObjects) DeleteObject(bucket, object string) error {
 func (a *azureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
-	// Azure doesn't return a unique upload ID and we use object name in place of it. Azure allows multiple uploads to
-	// co-exist as long as the user keeps the blocks uploaded (in block blobs) unique amongst concurrent upload attempts.
-	// Each concurrent client, keeps its own blockID list which it can commit.
-	uploadID = object
 	if metadata == nil {
-		// Store an empty map as a placeholder else ListObjectParts/PutObjectPart will not work properly.
 		metadata = make(map[string]string)
-	} else {
-		metadata = s3ToAzureHeaders(metadata)
 	}
-	a.metaInfo.set(uploadID, metadata)
-	return uploadID, nil
-}
 
-// CopyObjectPart - Not implemented.
-func (a *azureObjects) CopyObjectPart(srcBucket, srcObject, destBucket, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
-	return info, traceError(NotImplemented{})
-}
+	uploadID = mustGetAzureUploadID().String()
+	metadataObject := fmt.Sprintf(metadataObjectTemplate, uploadID)
+	metadata = s3ToAzureHeaders(metadata)
 
-// Encode partID, subPartNumber and md5Hex to blockID.
-func azureGetBlockID(partID, subPartNumber int, md5Hex string) string {
-	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%05d.%02d.%s", partID, subPartNumber, md5Hex)))
-}
-
-// Parse blockID into partID, subPartNumber and md5Hex.
-func azureParseBlockID(blockID string) (partID, subPartNumber int, md5Hex string, err error) {
-	var blockIDBytes []byte
-	if blockIDBytes, err = base64.StdEncoding.DecodeString(blockID); err != nil {
-		return
+	var jsonData []byte
+	if jsonData, err = json.Marshal(multipartMetadata{Name: object, Metadata: metadata}); err != nil {
+		return "", err
 	}
 
-	if _, err = fmt.Sscanf(string(blockIDBytes), "%05d.%02d.%s", &partID, &subPartNumber, &md5Hex); err != nil {
-		err = fmt.Errorf("invalid block id '%s'", string(blockIDBytes))
+	err = a.client.CreateBlockBlobFromReader(bucket, metadataObject, uint64(len(jsonData)), bytes.NewBuffer(jsonData), nil)
+	if err != nil {
+		return "", err
 	}
 
-	return
+	return uploadID, nil
+}
+
+// CopyObjectPart - Not implemented.
+func (a *azureObjects) CopyObjectPart(srcBucket, srcObject, destBucket, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
+	return info, traceError(NotImplemented{})
 }
 
 // PutObjectPart - Use Azure equivalent PutBlockWithLength.
 func (a *azureObjects) PutObjectPart(bucket, object, uploadID string, partID int, size int64, data io.Reader, md5Hex string, sha256sum string) (info PartInfo, err error) {
-	if meta := a.metaInfo.get(uploadID); meta == nil {
+	metadataObject := fmt.Sprintf(metadataObjectTemplate, uploadID)
+	if _, err = a.client.GetBlobMetadata(bucket, metadataObject); err != nil {
+		errorIf(err, "unable to read metadata object for upload ID %s", uploadID)
```

Comment:
`errorIf` statements can be removed as all the logging happens at the S3 layer

Reply:
I can't understand it. As `return info, traceError(InvalidUploadID{})` is done, I am unable to understand the problem of using `errorIf()`

Reply:
Convention that we follow is we don't log twice for the same error. We just log once at the S3 layer. Doing traceError() would mean we record the line number where it happened and hence know what the error is about.

If we are unable to return traceError for some reason then we log using errorIf